### PR TITLE
Add QNX/nto support

### PR DESCRIPTION
**Pull request by flba-eb**
This patch adds support to the QNX/nto operating system. This target is supported by Rust: https://github.com/rust-lang/rust/pull/102701.

cc @gh-tr

### CODE REVIEW

**Comment by newpavlov** [comment]
Is it possible to add a build job to our CI for this target?

File: src/util_libc.rs [hunk 1]
```diff
@@ -26,6 +26,8 @@ cfg_if! {
         use libc::__error as errno_location;
     } else if #[cfg(target_os = "haiku")] {
         use libc::_errnop as errno_location;
+    } else if #[cfg(target_os = "nto")] {
+        use libc::__get_errno_ptr as errno_location;
```
Comment (josephlr) [inline src/util_libc.rs:30]: Out of curiosity, where is this documented?

**Review by josephlr — CHANGES_REQUESTED** [review]
You can add this target to our CI by adding `x86_64-pc-nto-qnx7.1.0` after [this line](https://github.com/rust-random/getrandom/blob/759a9d743caa4d865a11972ed9882af2faf4eb3d/.github/workflows/tests.yml#L296).

File: src/lib.rs [hunk 2]
```diff
@@ -33,6 +33,7 @@
 //! | Web Browser and Node.js | `wasm*‑*‑unknown` | [`Crypto.getRandomValues`] if available, then [`crypto.randomFillSync`] if on Node.js, see [WebAssembly support]
 //! | SOLID             | `*-kmc-solid_*`    | `SOLID_RNG_SampleRandomBytes`
 //! | Nintendo 3DS      | `armv6k-nintendo-3ds` | [`getrandom`][1]
+//! | QNX Neutrino      | `*‑nto-qnx*`          | `/dev/urandom` [Random on QNX][14]
```
[the remark this answers is not in the document]
Comment (flba-eb) [reply src/lib.rs]: Done

File: src/use_file.rs [hunk 3]
```diff
@@ -30,7 +30,8 @@ const FILE_PATH: &str = "/dev/random\0";
     target_os = "redox",
     target_os = "dragonfly",
     target_os = "haiku",
-    target_os = "macos"
+    target_os = "macos",
+    target_os = "nto",
```
[the remark this answers is not in the document]
Comment (flba-eb) [reply src/use_file.rs:34]: Done, thanks for catching it!

File: src/util_libc.rs (hunk 1 above)
Comment (flba-eb) [reply src/util_libc.rs:30]: AFAIK this is not documented but `errno.h` of the QNX/nto toolchain uses it for `errno` as well:
```c
# define errno (*__get_errno_ptr())
```

**Comment by flba-eb** [comment]
> Is it possible to add a build job to our CI for this target?

I would love to have QNX/nto builds in CI, but I think it is not feasible right now because a license key is required. Also, the license seems not to work well in containers but works best with real hardware. I will forward this question to check if there is a possibility.

**Comment by josephlr** [comment]
> > Is it possible to add a build job to our CI for this target?
> 
> I would love to have QNX/nto builds in CI, but I think it is not feasible right now because a license key is required. Also, the license seems not to work well in containers but works best with real hardware. I will forward this question to check if there is a possibility.

That makes sense, we don't need to run the code, The line I suggested adding would only build the code, not execute it. Obviously actually running on qnx is better, but just building in our CI ensures that we don't inadvertently break things.

**Comment by josephlr** [comment]
> Not using the official QNX/nto toolchain would be tricky, because a special linker is used:

I realized that we cannot link without access to an OS specific toolchain. This is not the only target where we have this issue. That is why the line I suggested adding just compiles the code, without linking or running it.

**Comment by flba-eb** [comment]
I tested a bit with it but removed the test code again. We don't have (upstreamed) libc support yet (work in progress), which is why even `cargo check` does not work yet.

**Comment by josephlr** [comment]
> I tested a bit with it but removed the test code again. We don't have (upstreamed) libc support yet (work in progress), which is why even `cargo check` does not work yet.

That's good to know, and one of the reasons we like to build code for new targets. `cargo build` and `cargo check` will only compile the rust code, it shouldn't need to link anything. 

@newpavlov would you be okay accepting code that depends on not-yet-upstreamed libc support? Or should we just wait for the upstream libc changes before merging this?

**Comment by newpavlov** [comment]
@josephlr 
I think it's better to wait. If someone needs the support right away, then they simply can temporarily patch `getrandom` dependency in addition to `libc`.

**Comment by josephlr** [comment]
> @josephlr I think it's better to wait. If someone needs the support right away, then they simply can temporarily patch `getrandom` dependency in addition to `libc`.

Sounds reasonable.

@flba-eb feel free to ping this issue once the `libc` support for QNX/nto gets merged. Then I'd be happy to merge this.

**Comment by flba-eb** [comment]
@josephlr Ping! QNX support has been added to libc: https://github.com/rust-lang/libc/pull/3038

**Comment by josephlr** [comment]
> @josephlr Ping! QNX support has been added to libc: [rust-lang/libc#3038](https://github.com/rust-lang/libc/pull/3038)

Awesome!! Can you bump the `libc` version the `Cargo.toml` file to include the release that has that PR, and add `aarch64-unknown-nto-qnx710` to `tests.yml` [here](https://github.com/rust-random/getrandom/blob/759a9d743caa4d865a11972ed9882af2faf4eb3d/.github/workflows/tests.yml#L292)

**Comment by flba-eb** [comment]
@josephlr , done, please check. Check runs locally.

**Comment by josephlr** [comment]
The QNX tests are passing so I'm going to merge this. I'll fix the other tests in a followup PR.

**Review by josephlr — APPROVED** [review]

**Pull request merged**